pennylane/transforms/combine_global_phases.py

```diff
@@ -0,0 +1,73 @@
+"""
+Provides a transform to combine all ``qml.GlobalPhase`` gates in a circuit into a single one applied at the end.
+"""
+
+import pennylane as qml
+from pennylane.tape import QuantumScript, QuantumScriptBatch
+from pennylane.transforms import transform
+from pennylane.typing import PostprocessingFn
+
+
+@transform
+def combine_global_phases(tape: QuantumScript) -> tuple[QuantumScriptBatch, PostprocessingFn]:
+    """Combine all ``qml.GlobalPhase`` gates into a single ``qml.GlobalPhase`` operation.
+
+    Circuits passed through this transform will be updated to include the same operations and
+    measurements of the original input other than ``qml.GlobalPhase``.
+    A single ``qml.GlobalPhase`` operation will be added at the end of the new output circuit.
+    The total global phase is computed as the algebraic sum of all the phases in the original circuit.
+
+    Args:
+        tape (QNode or QuantumTape or Callable): the input circuit to be transformed.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumTape], function]:
+        the transformed circuit as described in :func:`qml.transform <pennylane.transform>`.
+
+    **Example**
+
+    Suppose we want to combine all the global phase gates in a given quantum circuit.
+    The ``combine_global_phases`` transform can be used to do this as follows:
+
+    .. code-block:: python3
+
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        @qml.transforms.combine_global_phases
+        def circuit():
+            qml.GlobalPhase(0.3, wires=0)
+            qml.PauliY(wires=0)
+            qml.Hadamard(wires=1)
+            qml.CNOT(wires=(1,2))
+            qml.GlobalPhase(0.46, wires=2)
+            return qml.expval(qml.X(0) @ qml.Z(1))
+
+    To check the result, let's print out the circuit:
+
+    >>> print(qml.draw(circuit)())
+    0: ──Y─────GlobalPhase(0.76)─┤ ╭<X@Z>
+    1: ──H─╭●──GlobalPhase(0.76)─┤ ╰<X@Z>
+    2: ────╰X──GlobalPhase(0.76)─┤
+    """
+
+    phi = 0
+    operations = []
+    for op in tape.operations:
+        if isinstance(op, qml.GlobalPhase):
+            phi += op.parameters[0]
+        else:
+            operations.append(op)
+
+    if phi != 0:
```

Comment:
```suggestion
    if not qml.math.allclose(phi, 0):
```
Sometimes it may not be exactly 0.

Reply:
We may have to also think about this in the context of jitting.

We try and make it so that all our code is compatible with `jax.jit` to the extent possible. And `jax.jit` is not a fan of any kind of comparison, as it's based around "abstract values" that can't be compared yet.

So we may actually need to do:
```suggestion
    if qml.math.is_abstract(phi) or not qml.math.allclose(phi, 0):
```

If `phi` is a jax tracer, we will need to apply the `GlobalPhase`, and we won't be able to use it in `qml.math.allclose`.

Reply:
@SimoneGasperini the following usage of `is_abstract` is needed here so that we don't perform boolean operations using a potentially abstract value, which can happen when using `jax.jit`.

```suggestion
    if not qml.math.is_abstract(phi) and not qml.math.allclose(phi, 0):
```

@astralcai I'd like to open using `==` vs `allclose` for discussion. I agree with `allclose` due to floating point precision, however, using `allclose` picks a default precision without any control given to the user. I'm personally hesitant to make a decision about which is better. A couple of potential ideas:
* Add `remove_zero_global_phase`, `atol` and `rtol` arguments to the transform. `remove_zero_global_phase` would be a boolean, and depending on whether it is `True` or `False`, we decide whether or not to remove the final global phase op if its phase is zero. `atol` and `rtol` would be given directly to `allclose`.
* We remove the validation about zero global phase altogether. I'd argue that the overhead of having one extra operation in the circuit is negligible, and we could potentially worry about it later.

Happy to hear thoughts 😄

Reply:
Potentially we include a final `GlobalPhase` if and only if the original circuit contains `GlobalPhase`, regardless of whether the phases cancel out with each other?

Reply:
@astralcai From my point of view, especially as a potential user of the transform, the solution you are proposing looks like the most reasonable here: if there is no `GlobalPhase` at all in the circuit, return it as it is; if there is one or more `GlobalPhase` gates in the circuit, always append a single `GlobalPhase` operation at the end even if all the phases sum up to 0.